File: backend/src/main/java/com/brassbook/service/ProfileService.java
package com.brassbook.service;

import com.brassbook.dto.request.ChangePasswordRequest;
import com.brassbook.dto.request.UpdateProfileRequest;
import com.brassbook.dto.response.ProfileResponse;
import com.brassbook.entity.User;
import com.brassbook.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
public class ProfileService {
    private final PasswordEncoder passwordEncoder;

    private final UserRepository userRepository;

    @Transactional(readOnly = true)
    public ProfileResponse getProfile(Long userId) {
        User user = findUserOrThrow(userId);
        return toProfileResponse(user);
    }

    @Transactional
    public void changePassword(Long userId, ChangePasswordRequest request) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("Пользователь не найден"));

        if (!passwordEncoder.matches(request.getCurrentPassword(), user.getPassword())) {
            throw new RuntimeException("Неверный текущий пароль");
        }

        user.setPassword(passwordEncoder.encode(request.getNewPassword()));
        userRepository.save(user);
    }

    @Transactional
    public ProfileResponse updateProfile(Long userId, UpdateProfileRequest request) {
        User user = findUserOrThrow(userId);

        if (request.getDisplayName() != null) {
            user.setDisplayName(request.getDisplayName());
        }

        if (request.getDisplaySurname() != null) {
            user.setDisplaySurname(request.getDisplaySurname());
        }

        if (request.getEmail() != null) {
            user.setEmail(request.getEmail());
        }

        User savedUser = userRepository.save(user);
        return toProfileResponse(savedUser);
    }

    @Transactional
    public void deleteAccount(Long userId) {
        User user = findUserOrThrow(userId);
        userRepository.delete(user);
    }

    private User findUserOrThrow(Long userId) {
        return userRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("Пользователь не найден"));
    }

    private ProfileResponse toProfileResponse(User user) {
        return ProfileResponse.builder()
                .id(user.getId())
                .email(user.getEmail())
                .displayName(user.getDisplayName())
                .displaySurname(user.getDisplaySurname())
                .photoUrl(user.getPhotoUrl())
                .role(user.getRole().name())
                .companyName(user.getCompanyName())
                .profession(user.getProfession())
                .inn(user.getInn())
                .createdAt(user.getCreatedAt())
                .build();
    }
}

File: frontend/services/ProfileService.ts
// services/ProfileService.ts
import { AxiosResponse } from "axios";
import $api from "../http";

export interface ProfileResponse {
    id: number;
    email: string;
    displayName?: string;
    displaySurname?: string;
    photoUrl?: string;
    role: string;
    createdAt: string;
    companyName?: string;
    profession?: string;
    inn?: number;
}

export interface UpdateProfileRequest {
    displayName?: string;
    displaySurname?: string;
    email?: string;
    companyName?: string;
    profession?: string;
    inn?: number;
}

export interface ChangePasswordRequest {
    currentPassword: string;
    newPassword: string;
}

export default class ProfileService {
    // GET /api/v1/profile
    static async getProfile(): Promise<AxiosResponse<ProfileResponse>> {
        return $api.get<ProfileResponse>('/profile');
    }

    // PUT /api/v1/profile
    static async updateProfile(data: UpdateProfileRequest): Promise<AxiosResponse<ProfileResponse>> {
        return $api.put<ProfileResponse>('/profile', data);
    }

    // PUT /api/v1/profile/password — эндпоинт ещё не готов, скажи напарнику
    static async changePassword(data: ChangePasswordRequest): Promise<AxiosResponse<void>> {
        return $api.put<void>('/profile/password', data);
    }

    // DELETE /api/v1/profile
    static async deleteAccount(): Promise<AxiosResponse<void>> {
        return $api.delete<void>('/profile');
    }

    // PUT /api/v1/profile/photo — multipart/form-data
    // Эндпоинта пока нет на бэке — заглушка готова к подключению
    static uploadPhoto(file: File): Promise<AxiosResponse<ProfileResponse>> {
        const formData = new FormData();
        formData.append("photo", file);
        return $api.put<ProfileResponse>("/profile/photo", formData, {
            headers: { "Content-Type": "multipart/form-data" },
        });
    }
}